## Dear Students,

Use the code available in Multiple Linear Regression Jupyter notebook under Week 2  in Canvas to complete data science assignment 03. Upload your assignment as a html file with the name assignment03_name_lastname.html

In [3]:
import pandas as pd
df = pd.read_csv("datasets/dataset_hw03.csv")
df.head()

Age        Income  Hours_Worked_Per_Week  Years_of_Experience  \
0   56  38062.928092              28.300832            28.783734   
1   46  17231.081305              11.277378            16.259079   
2   32  24356.921121              39.724851             7.584880   
3   60  18702.931453              57.722516            21.401708   
4   25  17798.041779              56.612592             4.080119   

   Job_Satisfaction_Score  Credit_Score  
0                     4.0    722.131044  
1                     3.0    609.546467  
2                     5.0    712.093586  
3                     5.0    678.674131  
4                     5.0    691.449956

Correlation Matrix:
                             Age  Hours_Worked_Per_Week  Years_of_Experience  \
Age                     1.000000               0.005158             0.936154   
Hours_Worked_Per_Week   0.005158               1.000000             0.008720   
Years_of_Experience     0.936154               0.008720             1.000000   
Job_Satisfaction_Score -0.017527               0.365242            -0.018052   
Credit_Score            0.171890               0.029851             0.175504   

                        Job_Satisfaction_Score  Credit_Score  
Age                                  -0.017527      0.171890  
Hours_Worked_Per_Week                 0.365242      0.029851  
Years_of_Experience                  -0.018052      0.175504  
Job_Satisfaction_Score                1.000000      0.029942  
Credit_Score                          0.029942      1.000000  


In [9]:
import numpy as np
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor
import pandas as pd

correlation_matrix = df.drop(columns='Income').corr()


X = df.drop(columns="Income") 
X_with_constant = sm.add_constant(X)  

vif_data = pd.DataFrame()
vif_data["Feature"] = X.columns
vif_data["VIF"] = [variance_inflation_factor(X_with_constant.values, i+1) for i in range(len(X.columns))]
print("Correlation Matrix:")
print(correlation_matrix)

print("\nVIF Data:")
print(vif_data)


Correlation Matrix:
                             Age  Hours_Worked_Per_Week  Years_of_Experience  \
Age                     1.000000               0.005158             0.936154   
Hours_Worked_Per_Week   0.005158               1.000000             0.008720   
Years_of_Experience     0.936154               0.008720             1.000000   
Job_Satisfaction_Score -0.017527               0.365242            -0.018052   
Credit_Score            0.171890               0.029851             0.175504   

                        Job_Satisfaction_Score  Credit_Score  
Age                                  -0.017527      0.171890  
Hours_Worked_Per_Week                 0.365242      0.029851  
Years_of_Experience                  -0.018052      0.175504  
Job_Satisfaction_Score                1.000000      0.029942  
Credit_Score                          0.029942      1.000000  

VIF Data:
                  Feature       VIF
0                     Age  8.094214
1   Hours_Worked_Per_Week  1.154703
2 

There is strong multicollinearity between Age and Years of Experience, which could affect the reliability of the regression model's coefficients

## Part 3: Build the Multiple Linear Regression Model

Now, I will build a multiple linear regression model using Income as the dependent variable and the other variables as independent predictors. I'll fit the model and display its statistical summary

In [12]:
y = df['Income']  
X = df.drop(columns='Income') 
X_with_constant = sm.add_constant(X)
#Regression MOdel
model = sm.OLS(y, X_with_constant).fit()
model_summary = model.summary()
model_summary


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Income   R-squared:                       0.346
Model:                            OLS   Adj. R-squared:                  0.343
Method:                 Least Squares   F-statistic:                     105.3
Date:                Tue, 29 Oct 2024   Prob (F-statistic):           3.03e-89
Time:                        17:12:36   Log-Likelihood:                -10606.
No. Observations:                1000   AIC:                         2.122e+04
Df Residuals:                     994   BIC:                         2.125e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                  -1.826e+04   4738.272     -3.854      0.000   -2.76e+04   -8963.367
Age                      438.0180     65.337      6.704      0.000     309.805     566.231
Hours_Worked_Per_Week    -11.3464     34.129     -0.332      0.740     -78.319      55.626
Years_of_Experience       59.8043     82.709      0.723      0.470    -102.500     222.109
Job_Satisfaction_Score  -944.0331    541.524     -1.743      0.082   -2006.694     118.628
Credit_Score              36.9833      6.230      5.937      0.000      24.759      49.208
==============================================================================
Omnibus:                        0.758   Durbin-Watson:                   1.984
Prob(Omnibus):                  0.684   Jarque-Bera (JB):                0.824
Skew:                           0.006   Prob(JB):                        0.662
Kurtosis:                       2.860   Cond. No.                     1.03e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.03e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
R-squared: 0.346
Adjusted R-squared: 0.343
F-statistic: 105.3 (p-value: 3.03e-89)

## Part 4: Questions

a) Are there correlated predictors?
VIF: Age and Years of Experience have high VIF values above 5, indicating potential multicollinearity.

b) Are the coefficients and standard errors reliable?
The coefficients for Age and Credit Score are statistically significant (p-values < 0.05), while Hours Worked Per Week, Years of Experience, and Job Satisfaction Score do not show strong significance based on their p-values.

c) How well does the model fit the data?
The R-squared of 0.346 indicates that about 34.6% of the variability in Income is explained by the predictors. 

d) Which predictors have the strongest relationship with income?
Age and Credit Score have the most significant impact on income based on their coefficients and p-values.